docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,698 @@
+
+However, the default implementation of non-interative defaults is within the scope of this ADR. Whatever design we use ultimately needs to support not using the non-interactive defaults. Meaning we should be able to encode block data into a square even if the messages are not arranged according to the non-interactive defaults. Again, this does not change the requirement that all share commitments signed over in PFDs consist only of subtree roots.
+
+## Detailed Design
+
+To recap the default constraints of arranging a square:
+
+- All messages must be ordered lexicographically by namespace.
+- The commitments signed over in each `MsgPayForData` must consist only of subtree roots of the data square.
+- If a `MsgPayForData` is added to the square, then its corresponding message must also be included.
+- There must not be a message without a `MsgPayForData` (does this need to be a rule? cc @adlerjohn).
+- Transactions with higher fees should be prioritized by default.
+- Ideally, the square should be filled as optimally as possible.
+
+For squares that are smaller than the max squaresize, the exact approach is much less important. This is because if we can't fit all of the transactions in a square, then by default we shouldn't be using that square size in the first place.
+
+Arranging the messages in the block to maximize for fees and optimize square space is an NP-hard problem. While the actual computation of the number of shares could be cached to an extent, each change to the square potentially affects the rest of the messages in the square. The only way to know for certain is to calculate the number of shares used. Calculating the exact number of bytes used is further complicated by the order of the steps due and our rampant use of varints, both in our encoding scheme and protobuf's. The example below shows how removing a single share (from the transactions in this case) could change the rest of the square and allow for a message that otherwise would not fit.
+
+![extra message that can't fit](./assets/extra-message-that-doesnt-fit.png "extra message")
+![fit the extra message by removing a tx](./assets/fit-extra-msg-byremoving-tx.png "extra message")
+
+To meet the above constraints, there are multiple refactors required.
+
+- Add metadata to wrapped transactions, connecting that transaction with a message that it pays for.
+- Refactor share splitting and merging to use the above metadata when decoding and encoding the square.
+- Implement the non-interactive default logic.
+- Implement the ability to traverse nmt tree to find subtree roots.
```

Comment:
Is there an issue about this?

Reply:
good point, I'll add links

we have some listed here https://github.com/celestiaorg/celestia-app/issues/382#issuecomment-1218372206

Reply:
703384a67328574b50c5cd7e4e0dd88367106390